qpid-jms-client/src/main/java/org/apache/qpid/jms/sasl/GssapiMechanism.java

```diff
@@ -0,0 +1,190 @@
+            }
+            loginContext.login();
+            subject = loginContext.getSubject();
+
+            return Subject.doAs(subject, new PrivilegedExceptionAction<byte[]>() {
+
+                @Override
+                public byte[] run() throws Exception {
+                    saslClient = Sasl.createSaslClient(new String[]{NAME}, null, protocol, serverName, null, null);
+                    if (saslClient.hasInitialResponse()) {
+                        return saslClient.evaluateChallenge(new byte[0]);
+                    }
+                    return null;
+                }
+            });
+        } catch (Exception e) {
+            throw new SaslException(e.toString(), e);
+        }
+    }
+
+    @Override
+    public byte[] getChallengeResponse(final byte[] challenge) throws SaslException {
+        try {
+            return Subject.doAs(subject, new PrivilegedExceptionAction<byte[]>() {
+                @Override
+                public byte[] run() throws Exception {
+                    return saslClient.evaluateChallenge(challenge);
+                }
+            });
+        } catch (PrivilegedActionException e) {
+            throw new SaslException(e.toString(), e);
+        }
+    }
+
+    @Override
+    public void verifyCompletion() throws SaslException {
+        boolean result = saslClient.isComplete();
+        saslClient.dispose();
+        if (!result) {
+            throw new SaslException("not complete");
+        }
+    }
+
+
+    @Override
+    public boolean isApplicable(String username, String password, Principal localPrincipal) {
+        return true;
+    }
+
+    private static final boolean IBM_JAVA =  System.getProperty("java.vendor").contains("IBM");
+    public static Configuration kerb5InlineConfig(String principal, boolean initiator) {
+        final Map<String, String> options = new HashMap<>();
+        if (IBM_JAVA) {
+            options.put("principal", principal);
+            options.put("renewable", "true");
+            options.put("credsType", initiator ? "initiator" : "acceptor");
+        } else {
+            options.put("isInitiator", String.valueOf(initiator));
+            options.put("principal", principal);
+            options.put("useKeyTab", "true");
+            options.put("storeKey", "true");
+            options.put("doNotPrompt", "true");
+            options.put("renewTGT", "true");
+            options.put("refreshKrb5Config", "true");
+            options.put("useTicketCache", "true");
+            String ticketCache = System.getenv("KRB5CCNAME");
+
+            if (IBM_JAVA) {
+                // If cache is specified via env variable, it takes priority
+                if (ticketCache != null) {
+                    // IBM JAVA only respects system property so copy ticket cache to system property
+                    // The first value searched when "useDefaultCcache" is true.
+                    System.setProperty("KRB5CCNAME", ticketCache);
```

Comment:
Those conditional have been replaced with the option to provide sasl.krb5.x overrides or additions as url parameters and if you wish fully configure the krb5 login module. 
The default options work for a client on the sun jdk, the ibm jdk needs url params.